### DISCUSSION TITLE
Keywords - Replace whitespaces with CamelCase

### DISCUSSION THREAD

**Discussion by dderom** [retorquere/zotero-better-bibtex#2367] [Q&A] [answered]
I use the Better BibTeX to import refs into Obsidian. I update my Zotero on a regular basis and add keywords as I go along. However, keywords with a whitespace (eg Natural Language Processing) are not allowed in Obsidian. One either has to use CamelCase, underscore or dash (aka NaturalLanguageProcessing or Natural_Language Processing or Natural-Language-Processing). 

I would like to convert on export the tags with Camelcase tags. I assume the Postscript on the export tab would be of use. However, can't seem to get it to work. 

Any pointers on where I could find some info on how to get this going?

**Comment by retorquere** [comment, marked as answer]
with a postscript:

```
if (Translator.BetterTeX && zotero.tags) {
  const tags = zotero.tags
    .map(tag => (typeof tag === 'string' ? tag : tag.tag))
    .map(tag => tag.replace(/\b\w/g, letter => letter.toUpperCase()).replace(/\s/g, ''))
  tex.add({ name: 'keywords', value: tags, enc: 'tags' })
}
```

**Reply by dderom** [reply]
Brilliant, that did it. Thanks for the swift reply ánd the plugin.

**Comment by Gladbjerg** [comment]
Pardon my question, but how do you get the keywords to export from Zotero? (I can't get it to work).

**Reply by retorquere** [reply]
To bib(la)tex? That should happen by default.

**Reply by Gladbjerg** [reply]
Yes, to the JSON file. Mine doesn't. I would not rule out, that I have done something at one point, when I follow a setup guide (for automatic import into Obsidian.

**Reply by Gladbjerg** [reply]
Is it possible to reset the settings (so it's like the first time it's installed)

**Reply by retorquere** [reply]
To BBT JSON? That has tags, and there isn't anything configurable about that.

**Reply by Gladbjerg** [reply]
I followed a guide to it.. 

https://www.youtube.com/watch?v=D9ivU_IKO6M&ab_channel=ArtemKirsanov

I have opened the JSON file in Bbedit, and it does not contain tags.

**Reply by retorquere** [reply]
> Is it possible to reset the settings (so it's like the first time it's installed)

If you save this to a file you can import it in BBT prefs under the import tab

```
{"config":{"preferences":{"ascii":"","asciiBibLaTeX":false,"asciiBibTeX":true,"autoAbbrev":false,"autoAbbrevStyle":"","autoExport":"immediate","autoExportDelay":5,"autoExportIdleWait":10,"autoExportPathReplaceDiacritics":false,"autoExportPathReplaceDirSep":"-","autoExportPathReplaceSpace":" ","automaticTags":true,"autoPinDelay":0,"auxImport":false,"baseAttachmentPath":"","biblatexExtendedDateFormat":true,"biblatexExtendedNameFormat":false,"biblatexExtractEprint":true,"bibtexParticleNoOp":false,"bibtexURL":"off","cache":true,"cacheFlushInterval":5,"charmap":"","citeCommand":"cite","citekeyFold":true,"citekeyFormat":"auth.lower + shorttitle(3,3) + year","citekeyFormatBackup":"","citekeySearch":true,"citekeyUnsafeChars":"\\\"#%'(),={}~","csquotes":"","DOIandURL":"both","exportBibTeXStrings":"off","exportBraceProtection":true,"exportTitleCase":true,"extraMergeCitekeys":false,"extraMergeCSL":false,"extraMergeTeX":false,"git":"config","import":true,"importBibTeXStrings":true,"importCaseProtection":"as-needed","importCitationKey":true,"importExtra":true,"importJabRefAbbreviations":true,"importJabRefStrings":true,"importNoteToExtra":"","importSentenceCase":"on+guess","importUnknownTexCommand":"ignore","itemObserverDelay":5,"jabrefFormat":0,"jieba":false,"keyConflictPolicy":"keep","keyScope":"library","kuroshiro":false,"language":"langid","mapMath":"","mapText":"","mapUnicode":"conservative","parseParticles":true,"patchDates":"dateadded=dateAdded, date-added=dateAdded, datemodified=dateModified, date-modified=dateModified","platform":"","postscript":"","postscriptOverride":"","preferencesOverride":"","qualityReport":false,"quickCopyEta":"","quickCopyMode":"latex","quickCopyOrgMode":"zotero","quickCopyPandocBrackets":false,"quickCopySelectLink":"zotero","rawImports":false,"rawLaTag":"#LaTeX","relativeFilePaths":false,"retainCache":false,"scrubDatabase":false,"separatorList":"and","separatorNames":"and","skipFields":"","skipWords":"a,ab,aboard,about,above,across,after,against,al,along,amid,among,an,and,anti,around,as,at,before,behind,below,beneath,beside,besides,between,beyond,but,by,d,da,das,de,del,dell,dello,dei,degli,della,dell,delle,dem,den,der,des,despite,die,do,down,du,during,ein,eine,einem,einen,einer,eines,el,en,et,except,for,from,gli,i,il,in,inside,into,is,l,la,las,le,les,like,lo,los,near,nor,of,off,on,onto,or,over,past,per,plus,round,save,since,so,some,sur,than,the,through,to,toward,towards,un,una,unas,under,underneath,une,unlike,uno,unos,until,up,upon,versus,via,von,while,with,within,without,yet,zu,zum","startupProgress":"popup","strings":"","stringsOverride":"","testing":false,"verbatimFields":"url,doi,file,pdf,ids,eprint,/^verb[a-z]$/,groups,/^citeulike-linkout-[0-9]+$/, /^bdsk-url-[0-9]+$/","warnBulkModify":10,"warnTitleCased":false}}}
```

**Reply by retorquere** [reply]
> I followed a guide to it..
> 
> https://www.youtube.com/watch?v=D9ivU_IKO6M&ab_channel=ArtemKirsanov

I don't know what the relevant part is in the video, but you can just export any item(s) to `BetterBibTeX JSON`. You may be exporting to `CSL JSON`, which doesn't have the tags.

**Reply by Gladbjerg** [reply]
I am exporting it to CSL JSON… That would explain it. Thank you for clarifying. I will try to change it from CSL to BBT. 

And sorry for being a N00b here..

**Reply by Gladbjerg** [reply]
I am using the "obsidian-citation-plugin" for Obsidian, and that one needs CSL JSON.

@dderom Do you use a plugin for Obsidian?

**Reply by retorquere** [reply]
> I am exporting it to CSL JSON… That would explain it. Thank you for clarifying. I will try to change it from CSL to BBT.
> 
> And sorry for being a N00b here..

No worries. It's always OK to ask.

@dderom exports to bibtex. You could add the keywords using a postscript for CSL-JSON, but no citation processor is going to pick them up.

**Reply by Gladbjerg** [reply]
> > I am exporting it to CSL JSON… That would explain it. Thank you for clarifying. I will try to change it from CSL to BBT.
> > And sorry for being a N00b here..
> 
> No worries. It's always OK to ask.
> 
> @dderom exports to bibtex. You could add the keywords using a postscript for CSL-JSON, but no citation processor is going to pick them up.

I am looking at the documentation for the obsidian-citation-plugin" and it can handle CLS JSON or BBT.bib .. 
If I do a bbt.bib it does export the tags as "keywords".. I just don't know how to import that to Obsidian.

**Reply by retorquere** [reply]
this postscript will add the keywords to CSL exports, but I don't know how you'd use them.

```
if (Translator.BetterCSL && zotero.tags) {
  csl.keywords = zotero.tags.map(tag => (typeof tag === 'string' ? tag : tag.tag))
}
```

**Reply by Gladbjerg** [reply]
> this postscript will add the keywords to CSL exports, but I don't know how you'd use them.
> 
> ```
> if (Translator.BetterCSL && zotero.tags) {
>   csl.keywords = zotero.tags.map(tag => (typeof tag === 'string' ? tag : tag.tag))
> }
> ```

Thank you. Would that also remove the whitespace? (I prefer an underscore).

**Reply by Gladbjerg** [reply]
And right now I am stuck on not being able to import them to Obsidian (at least not using the mentioned plugin).

**Reply by retorquere** [reply]
> Thank you. Would that also remove the whitespace? (I prefer an underscore).

No, for that you'd need

```
if (Translator.BetterCSL && zotero.tags) {
  csl.keywords = zotero.tags.map(tag => (typeof tag === 'string' ? tag : tag.tag)).map(tag => tag.replace(/\s+/g, '_'))
}
```

> And right now I am stuck on not being able to import them to Obsidian (at least not using the mentioned plugin).

I can't help with that, I'm not a regular Obsidian user.

**Reply by dderom** [reply]
As soon as you have the zotero postscript, you need to add a template to import the  keywords. In other words, go to the settings/options of the plugin (cf. Citations) of Obsidian and add the following in the 'literature note content template':
```
---
type: reference
authors: {{authorString}}
year: {{year}}
DOI: {{DOI}}
URL: {{URL}}
year: {{year}}
citekey: {{citekey}}
---
# {{title}} ({{year}}_

## Abstract
{{abstract}}

## Tags
{{#each entry.data.fields.keywords}}{{#if @last}}#{{this}} {{else}}#{{this}} {{/if}}{{/each}}
```
Simply giving you a template I use, where the first part is front matter, the second imports the keywords of the zotero reference. So essentially, you need only the last part to import the keywords:
```
## Tags
{{#each entry.data.fields.keywords}}{{#if @last}}#{{this}} {{else}}#{{this}} {{/if}}{{/each}}
```

**Reply by Gladbjerg** [reply]
> As soon as you have the zotero postscript, you need to add a template to import the keywords. In other words, go to the settings/options of the plugin (cf. Citations) of Obsidian and add the following in the 'literature note content template':
> 
> ```
> ---
> type: reference
> authors: {{authorString}}
> year: {{year}}
> DOI: {{DOI}}
> URL: {{URL}}
> year: {{year}}
> citekey: {{citekey}}
> ---
> # {{title}} ({{year}}_
> 
> ## Abstract
> {{abstract}}
> 
> ## Tags
> {{#each entry.data.fields.keywords}}{{#if @last}}#{{this}} {{else}}#{{this}} {{/if}}{{/each}}
> ```
> 
> Simply giving you a template I use, where the first part is front matter, the second imports the keywords of the zotero reference. So essentially, you need only the last part to import the keywords:
> 
> ```
> ## Tags
> {{#each entry.data.fields.keywords}}{{#if @last}}#{{this}} {{else}}#{{this}} {{/if}}{{/each}}
> ```

Where did you find the documentation regarding importing the tags? I can see, that you import every tag as a tag with # before the tha (e.g. #tag1, #tag2). It is possbile to change the import, so it will be a comma e.g. Tag1, Tag2 etc.

**Comment by Gladbjerg** [comment]
So I have had a chance to play around with it.. I am now able to export the Tags to the BBT file. And also import it into Obsidian. But how do I get the tags to be seperated with a "," and not a "#" in front of a tag (because I don't want to use them with Obsidian keywords, but rather as YAML). eg. tags: tag1, tag2, tag3   and not tags: #tag1 #tag2 #tag3. If I do export via CamelCase, then it capitalises the letters after æ,ø,å eg. LæRebøGer where it should be Lærebøger (which it is in the Zotero Tag)

**Reply by retorquere** [reply]
I need to know what you mean by "BBT file" here, because the discussion above switches between BibLaTeX, CSL-JSON, and BBT-JSON. I am not adding `#` marks anywhere, so I don't know where those come from, or at what stage of the process you are following they appear.

The camelcase issue is because `\w` only understands ascii, so assuming you're exporting to BibTeX, try this:

```
if (Translator.BetterTeX && zotero.tags) {
  const tags = zotero.tags
    .map(tag => (typeof tag === 'string' ? tag : tag.tag))
    .map(tag => tag.replace(/(^|\s+)./g, letter => letter.toUpperCase()).replace(/\s/g, ''))
  tex.add({ name: 'keywords', value: tags, enc: 'tags' })
}
```

**Reply by Gladbjerg** [reply]
Sorry for not explaining it better. By BBT I mean the Better BibLaTeX file. 
NO I found out, that the adding of the # is done at import into Obsidian. And the problem with the formatting is also in the import from the Better BibLaTeX file.. 

My problem is when I import from the Better BibLaTeX file with the Obsidian Citation plugin..  

Sorry for the confusion.

**Reply by retorquere** [reply]
If the `#` characters are added after the .bib file is produced, there's nothing that BBT can do. You'll have to ask the author of the Obsidian Citation plugin.
